provider/apple.go

```diff
@@ -0,0 +1,526 @@
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("content-type", appleRequestContentType)
+	req.Header.Add("accept", AcceptJSONHeader)
+	req.Header.Add("user-agent", defaultUserAgent) // apple requires a user agent
+
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	// Trying to decode (unmarshal json) data of response
+	err = json.NewDecoder(res.Body).Decode(result)
+	if err != nil {
+		return errors.Wrap(err, "unmarshalling data from apple service response failed")
+	}
+
+	// If above operation done successfully checking a response code and error descriptions, if one exist.
+	// Apple service will response either 200 (OK) or 400 (any error).
+	if res.StatusCode != http.StatusOK || result.Error != "" {
+		return fmt.Errorf("apple token service error: %s", result.Error)
+	}
+
+	defer func() {
+		if err = res.Body.Close(); err != nil {
+			ah.L.Logf("[ERROR] close request body failed when get access token: %v", err)
+		}
+	}()
+
+	return err
+}
+
+// createClientSecret use for create the JWT client secret required to make requests to the Apple validation server.
+// for more details go to link: https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens#3262048
+func (ah *AppleHandler) createClientSecret() (string, error) {
+
+	if ah.conf.privateKey == nil {
+		return "", errors.New("private key can't be empty")
+	}
+	// Create a claims
+	now := time.Now()
+	exp := now.Add(time.Minute * 30).Unix() // default value
+
+	claims := &jwt.StandardClaims{
+		Issuer:    ah.conf.TeamID,
+		IssuedAt:  now.Unix(),
+		ExpiresAt: exp,
+		Audience:  "https://appleid.apple.com",
+		Subject:   ah.conf.ClientID,
+	}
+
+	tkn := jwt.NewWithClaims(jwt.SigningMethodES256, claims)
+	tkn.Header["alg"] = "ES256"
+	tkn.Header["kid"] = ah.conf.KeyID
+
+	return tkn.SignedString(ah.conf.privateKey)
+}
+
+func (ah *AppleHandler) parseUserData(user *token.User, jUser string) {
+
+	type UserData struct {
+		Name struct {
+			FirstName string `json:"firstName"`
+			LastName  string `json:"lastName"`
+		} `json:"name"`
+		Email string `json:"email"`
+	}
+
+	var userData UserData
+
+	// Catch error for log only. No need break flow if user name doesn't exist
+	if err := json.Unmarshal([]byte(jUser), &userData); err != nil {
+		ah.L.Logf("[DEBUG] failed to parse user data %s: %v", user, err)
+		user.Name = "noname_" + user.ID[6:12] // paste noname if user name failed to parse
+		return
+	}
+
+	user.Name = fmt.Sprintf("%s %s", userData.Name.FirstName, userData.Name.LastName)
+}
+
+func (ah *AppleHandler) prepareLoginURL(state, path string) (string, error) {
+
+	scopesList := strings.Join(ah.conf.scopes, " ")
+
+	authURL, err := url.Parse(ah.endpoint.AuthURL)
+	if err != nil {
+		return "", err
+	}
+
+	query := authURL.Query()
+	query.Set("state", state)
+	query.Set("response_type", "code")
+	query.Set("response_mode", "form_post")
```

Comment:
@zebox hi, I was wondering if there was any particular reason to hardcode `response_mode`.

Reply:
@akellbl4 hi, according to [documentation](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_js/incorporating_sign_in_with_apple_into_other_platforms#3332115) `response_mode` has three parameters: `query`, `fragment `, `form_post `. But `query` and `fragment` includes response parameters to the `redirectURI`. I prefer to avoid show explicit response parameters, with sensitive user data, in a browser URL. Therefore, I send it as form data.

It's like for `login` scheme with `GET` method, where credentials send at query URL, but I prefer send in body with `POST`. 
I know that credentials should be hashed at frontend side, before send to backend, but i think better hides this data from explicit view in a browser. I understand that doesn't added some strong security, but  it's hides credentials from access logs (which logs parameters of URL query on a backend side).